lib/Promises/Deferred.pm

```diff
@@ -4,186 +4,207 @@ package Promises::Deferred;
-
-    my $d = ( ref $self )->new;
+sub _invoke_cb {
+    my $cb= shift;
+    my $self= shift @$cb;
 
-    if (defined $callback) {
-        my ( @result, $method );
-        my $finish_d = sub { $d->$method(@result); () };
-
-        my $f = sub {
-            ( $method, @result ) = @_;
-            local $@;
-            my ($p) = eval { $callback->(@result) };
-            if ( $p && blessed $p && $p->can('then') ) {
-                return $p->then( $finish_d, $finish_d );
+    if (my $invoke_callback= $cb->[$$self->{state}]) {
+        eval {
+            my @result= $invoke_callback->(@{$$self->{result}});
+            if (my $next= $cb->[0]) {
+                if (@result == 1 && is_blessed_ref($result[0]) && $result[0]->can('then')) {
+                    if ($result[0]->isa("Promises::Promise")) {
+                        _replace_promise($next, ${$result[0]});
+                    } elsif ($result[0]->isa("Promises::Deferred")) {
+                        _replace_promise($next, $result[0]);
+                    } else {
+                        $result[0]->then(sub {
+                            $next->resolve(@_); ();
+                        }, sub {
+                            $next->reject(@_); ();
+                        });
+                    }
+                } else {
+                    $next->resolve(@result);
+                }
             }
-            $finish_d->();
-            ();
+            1;
+        } or do {
+            my $error= $@ || '';
+            $cb->[0]->reject($error) if $cb->[0];
         };
 
-        push @{ $self->{'resolved'} } => sub { $f->( 'resolve', @_ ) };
-        push @{ $self->{'rejected'} } => sub { $f->( 'reject',  @_ ) };
-
-        $self->_notify unless $self->is_in_progress;
+    } elsif ($cb->[0]) { # Passthrough
+        if ($$self->{state} == 1) {
+            $cb->[0]->resolve(@{$$self->{result}});
+        } else {
+            $cb->[0]->reject(@{$$self->{result}});
+        }
     }
-    $d->promise;
 
+    1;
 }
 
-sub _wrap {
-    my ( $self, $d, $f, $method ) = @_;
-
-    return sub { $d->$method( @{ $self->result } ) }
-        unless defined $f;
-
-    return sub {
-        local $@;
-        my ( @results, $error );
-        eval {
-            @results = do { $f->(@_) };
-            1;
-        }
-            || do { $error = $@ || 'Unknown reason' };
+# Optimization: when returning a promise from a callback,
+# we can replace the original promise with the new one.
+# After all, its results, state and callback will all
+# be equal to those of the newly returned promise.
+sub _replace_promise {
+    my ($orig_promise, $new_promise)= @_;
+    my $orig_inner= $$orig_promise;
+    my $new_inner= $$new_promise;
+
+    # Do the replacement
+    $$orig_promise= $$new_promise;
+
+    my $callbacks= delete $orig_inner->{cb};
+    if ($new_inner->{state}) {
+        _invoke_cbs($callbacks);
+    } else {
+        push @{$new_inner->{cb}}, @$callbacks;
+    }
```

Comment:
So I suspect this is the reason for the RefOfRef instance, ... not sure it is worth it given how ugly it is.